Add docs (#294)
CQCL/hugr-qir: docs/wasm.md

# WASM Support

`hugr-qir` supports lowering programs that use the wasm extension and refer to functions in a provided WASM module.

Currently, the only supported guppylang wasm platform is `WasmPlatform.H2`. This is not
the default and must be explicitly set as a parameter to the guppylang `@wasm_module` decorator.

> Please note that the guppylang function for spawning multiple wasm contexts (`guppylang.std.qsystem.spawn_wasm_contexts`) is not supported. Only a single wasm context is allowed, and it is generated by constructing the decorated wasm module class using the syntax `MyWasm(nat(1))` (see [Examples: WASM integration](examples/wasm.md)).

Reviewer: ```suggestion
Only a single WASM context is supported. `guppylang.std.qsystem.spawn_wasm_contexts` is not supported, and the context is created by constructing the decorated WASM module class with `MyWasm(nat(1))` (see [Examples: WASM integration](examples/wasm.md)).
```

> The only supported WASM function parameter types in the compiled wasm file are 32-bit integers. However, it is acceptable if a wasm file contains unsupported functions that are not called in the guppy program.

Reviewer: ```suggestion
In the compiled WASM file, the only supported function parameter type is a 32-bit integer. A WASM file may still contain unsupported functions, as long as the Guppy program does not call them.
```
